src/server/grpc_service.rs

```diff
@@ -0,0 +1,636 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::boxed::FnBox;
+use std::fmt::Debug;
+use std::io::Write;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use mio::Token;
+use grpc::{RpcContext, UnarySink, ClientStreamingSink, RequestStream};
+use futures::{future, Future, Stream};
+use futures::sync::oneshot;
+use tokio_core::reactor::Remote;
+use protobuf::RepeatedField;
+use kvproto::tikvpb_grpc;
+use kvproto::raft_serverpb::*;
+use kvproto::kvrpcpb::*;
+use kvproto::coprocessor::*;
+use kvproto::errorpb::{Error as RegionError, ServerIsBusy};
+
+use util::worker::Scheduler;
+use util::buf::PipeBuffer;
+use storage::{self, Storage, Key, Options, Mutation};
+use storage::txn::Error as TxnError;
+use storage::mvcc::Error as MvccError;
+use storage::engine::Error as EngineError;
+use super::transport::RaftStoreRouter;
+use super::coprocessor::{RequestTask, EndPointTask};
+use super::snap::Task as SnapTask;
+use super::metrics::*;
+use super::Error;
+
+#[derive(Clone)]
+pub struct Service<T: RaftStoreRouter + 'static> {
+    core: Remote,
+    // For handling KV requests.
+    storage: Storage,
+    // For handling coprocessor requests.
+    end_point_scheduler: Scheduler<EndPointTask>,
+    // For handling raft messages.
+    ch: T,
+    // For handling snapshot.
+    snap_scheduler: Scheduler<SnapTask>,
+    token: Arc<AtomicUsize>, // TODO: remove it.
+}
+
+impl<T: RaftStoreRouter + 'static> Service<T> {
+    pub fn new(core: Remote,
+               storage: Storage,
+               end_point_scheduler: Scheduler<EndPointTask>,
+               ch: T,
+               snap_scheduler: Scheduler<SnapTask>)
+               -> Service<T> {
+        Service {
+            core: core,
+            storage: storage,
+            end_point_scheduler: end_point_scheduler,
+            ch: ch,
+            snap_scheduler: snap_scheduler,
+            token: Arc::new(AtomicUsize::new(1)),
+        }
+    }
+}
+
+fn make_callback<T: Debug + Send + 'static>() -> (Box<FnBox(T) + Send>, oneshot::Receiver<T>) {
+    let (tx, rx) = oneshot::channel();
+    let callback = move |resp| {
+        tx.send(resp).unwrap();
+    };
+    (box callback, rx)
+}
+
+impl<T: RaftStoreRouter + 'static> tikvpb_grpc::Tikv for Service<T> {
+    fn kv_get(&self, _: RpcContext, mut req: GetRequest, sink: UnarySink<GetResponse>) {
+        let timer = GRPC_MSG_HISTOGRAM_VEC.with_label_values(&["kv_get"]).start_timer();
+
+        let storage = self.storage.clone();
+        self.core.spawn(move |_| {
+            let (cb, future) = make_callback();
+            storage.async_get(req.take_context(),
```

Comment:
Since async_XXX always return success, can we remove the return value, so we can avoid the unwrap call.